virttest/migration.py

```
@@ -271,26 +283,187 @@ def do_migration(self, vms, srcuri, desturi, migration_type,
+            To execute virsh event command to collect the domain events
+
+            :return: VirshSession to retrieve the events
+            """
+            cmd = "event --loop --all"
+            virsh_event_session = virsh.VirshSession(virsh_exec=virsh.VIRSH_EXEC,
+                                                     auto_close=True,
+                                                     uri=srcuri)
+            virsh_event_session.sendline(cmd)
+            logging.debug("Begin to collect domain events...")
+            return virsh_event_session
+
+        def _need_collect_events(funcs_to_run):
+            """
+            Check if there is a need to run a command to collect domain events
+
+            :param funcs_to_run: the functions to be run. It can be a list
+                                 or a single function
+                For example,
+                funcs_to_run = [{"func": <function check_established at 0x7f5833687510>,
+                                 "after_event": "iteration: '1'",
+                                 "before_event": "Suspended Migrated",
+                                 "before_pause": "yes"},
+                                {"func": <function domjobabort at 0x7f5835cd08c8>,
+                                 "before_pause": "no"}
+                               ]
+            :return: boolean, True to collect events, otherwise False
+            """
+            if not funcs_to_run:
+                return False
+            if isinstance(funcs_to_run, list):
+                for one_func in funcs_to_run:
+                    if isinstance(one_func, dict):
+                        after_event = one_func.get('after_event')
+                        before_event = one_func.get('before_event')
+                        if any([after_event, before_event]):
+                            return True
+                    else:
+                        # It should be a FunctionType
+                        return False
```

As we discussed, let's raise an exception here. Thanks!

---

Updated